ext/json/config.m4

```diff
@@ -1,19 +1,8 @@
-PHP_ARG_ENABLE([json],
-  [whether to enable JavaScript Object Serialization support],
-  [AS_HELP_STRING([--disable-json],
-    [Disable JavaScript Object Serialization support])],
-  [yes])
-
-if test "$PHP_JSON" != "no"; then
-  AC_DEFINE([HAVE_JSON],1 ,[whether to enable JavaScript Object Serialization support])
```

Comment:
I would keep the define. There may be other extensions (not bundled with PHP) which check for this. It costs essentially nothing.

Reply:
We didn't leave these for `ext/hash` and actively removed similar defines for other builtin extensions like `ext/pcre`, `ext/spl` about a year or so ago, so if this RFC passes I think it should follow the same style

Reply:
It makes things needlessly annoying for PECL extensions etc that need to support older versions. We kept around TSRMLS, we can keep this around. Not all precedents are good!

Reply:
> It makes things needlessly annoying for PECL extensions etc that need to support older versions. We kept around TSRMLS, we can keep this around. Not all precedents are good!

It also makes it obvious that the extension authors are able to use json functionality unconditionally, and discourages people from manually patching php 8 to disable json and have HAVE_JSON set to 0.

I'm having a hard time finding how often/where `HAVE_JSON` is actually used - I suppose downloading the top 1000 pecl releases is possible but haven't done that before. Then again, lack of recommendations or tutorials on how to use config.m4 may contribute to that.

https://github.com/php-memcached-dev/php-memcached/blob/v3.1.5/config.m4 (checks for the existence of the header)

https://github.com/mongodb/mongo-php-driver/blob/master/config.m4 uses `PHP_ADD_EXTENSION_DEP(mongodb, json)`

We made TSRMLS a no-op in 7.0 and removed it from 8.0, but it's much more frequently used than HAVE_JSON, so the numbers of lines of code affected is very different

Altogether, though - whether or not HAVE_JSON is there isn't very important to me, but having a clear policy on when to remove macros would help. I'll probably end up restoring the macro since multiple people have raised objections (and keep the rfc to a single vote).

Reply:
> Then again, lack of recommendations or tutorials on how to use config.m4 may contribute to that.

I checked if there were recommendations published since the last time I looked at a config.m4 with dependencies - I had similar questions back in 2017, and solved those for my use case by looking at what other extensions were doing in config.m4 and config.w32, which required a few revisions to fix on rarer build setups (e.g. statically compiled an extension on windows).

https://github.com/phpinternalsbook/PHP-Internals-Book/issues/43

Reply:
I think removing HAVE_SPL or some others too quicly was a bad idea, and will prefer to have HAVE_JSON defined in 8.

Reply:
I'm keeping the `HAVE_JSON` constant in config.m4 for non-Windows builds.

One thing I discovered is that config.w32 apparently never defined `HAVE_JSON` - this PR will not change that. I initially thought Windows builds were somehow different, but was mistaken

- In other extensions, AC_DEFINE is used to define those constants, e.g. https://github.com/php/php-src/blob/PHP-7.4.5/ext/zip/config.w32#L17 `AC_DEFINE('HAVE_ZIP', 1);`.

Reply:
I assume that `HAVE_JSON` has never been defined on Windows, because the macro is not used (at least in php-src).

Reply:
The Windows build environment used PHP_JSON in stead of HAVE_JSON. Since PHP_JSON has been removed now, extension maintainers can no longer check in their config.w32 if json is enabled. They just have to assume json is enabled in PHP 8. See https://github.com/php-ds/ext-ds/pull/162 for a way to do that.

Use something like this for a check in config.w32 on PHP 5 or 7:
```
var dll = get_define('PHPDLL');
if (null != dll.match(/^php[57]/)) {
    ADD_EXTENSION_DEP('extension name', 'json');
}
```

Reply:
See https://github.com/mongodb/mongo-php-driver/issues/1161#issuecomment-715508317 for the remark of the mongodb maintainers about their inability to check if the standard and date extensions are enabled. This is the same for json in PHP 8 now as well.

Reply:
Thanks @Jan-E for pointing this out. I think the proper fix would be:
````.diff
 ext/json/config.w32 | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/ext/json/config.w32 b/ext/json/config.w32
index 9d7881118d..255f278772 100644
--- a/ext/json/config.w32
+++ b/ext/json/config.w32
@@ -1,7 +1,7 @@
 // vim:ft=javascript
 
 EXTENSION('json', 'json.c', false /* never shared */, "/DZEND_ENABLE_STATIC_TSRMLS_CACHE=1");
-
+PHP_JSON="yes";
 ADD_SOURCES(configure_module_dirname, "json_encoder.c json_parser.tab.c json_scanner.c", "json");
 
 ADD_MAKEFILE_FRAGMENT();
````
I need to check phpize builds; if these are good, I'll commit that fix.

Reply:
Committed as 560970138915791fe29380c8f31a7ec18379b00f.

Reply:
If I apply this commit to PHP 8.0.0RC2, I am running into another Jscript error with mongodb in ext/mongodb:

> N:\php-sdk\php80dev\configure.js(8470, 3) Microsoft JScript runtime error: Path not found

Line 8470 in configure.js is the 7th line in the current config.w32 of mongodb:
https://github.com/mongodb/mongo-php-driver/blob/master/config.w32#L7
```
configure_module_dirname = condense_path(FSO.GetParentFolderName('N:\\php-sdk\\php80dev\\ext\\mongodb\\config.w32'));
// vim:ft=javascript

function mongodb_generate_header(inpath, outpath, replacements)
{
  STDOUT.WriteLine("Generating " + outpath);

  var infile = FSO.OpenTextFile(inpath, 1);
  var outdata = infile.ReadAll();
  infile.Close();

  for (var key in replacements) {
    var replacement = replacements[key];

    if (typeof replacement === 'string') {
      replacement = replacement.replace(/"/g, '\\"');
    }

    outdata = outdata.replace(new RegExp('@' + key + '@', 'g'), replacement);
  }

  var outfile = FSO.CreateTextFile(outpath, true);
  outfile.Write(outdata);
  outfile.Close();
}
```

Reply:
Forget my last comment for now. It was a missing libbson. The inpath was pointing to:

> mongodb/src/libmongoc/src/libbson/src/bson/bson-config.h.in

Reply:
With a properly installed ext/mongodb compilation runs into the expected errors, because the mongodb extension is not fit yet for PHP 8.

ext/ds with a `ADD_EXTENSION_DEP('ds', 'json')` compiles fine after https://github.com/php/php-src/commit/560970138915791fe29380c8f31a7ec18379b00f